Load tests lane configuration EVM2EVM

asoliman92 opened the pull request:
This PR adds lane configuration to load tests. Currently supporting only EVM2EVM.  

The configuration is used in crib:  https://github.com/smartcontractkit/crib/pull/577

Then In ccip_test we [discover](https://github.com/smartcontractkit/chainlink/pull/18155/files#diff-2fd8dcdd52e787346ae0d444ac60e6cd76af3e3658c5f104966ce3388338b494R295) lanes from existing setup and send messages based on the lane config. 

We have 2 modes, any-to-any and random. Each is implemented and generated in the LaneConfiguration.

Example config in crib will look like:
```
GETH_CHAINS_COUNT=5
DON_NODE_COUNT=4
# Lanes configuration
LANE_MODE="random-lanes"
NUM_LANES=11 # This will be 12 lanes as all lanes should be bidirectional
```

Bug fixes:
* Swapping src and dst in offRamp while connecting lanes in `ccip_deployer`
* Use separate sender for each src->dst lane. (was using only one sender per one source chain) in `ccip_test`

Optimizations:
* Lane deployment should be faster now after extracting onRamp/offRamp changesets to run outside the loop and parallelizing the other changesets.

asoliman92 committed:
Add lane configuration and include it in evm2evm lanes setup

asoliman92 committed:
small fixes to lane_config

asoliman92 committed:
Add Solana lanes based on config

asoliman92 committed:
Add discover lanes functionality to lane configuration ONLY EVM FOR NOW

Given a deployed environment construct lane configuration with the existing lanes

asoliman92 committed:
Use discover lane functionality in tests

asoliman92 committed:
extra logging

asoliman92 committed:
minimum lanes is at least each chain twice source and dest

asoliman92 committed:
Refactor destination gun to remove chainOffset and implement mutex for source chain locking

asoliman92 commented on deployment/environment/crib/ccip_deployer.go:
```diff
@@ -607,42 +621,113 @@ func setupSolLinkPools(e *cldf.Environment) (cldf.Environment, error) {
+	return false
+}
+
+func setupSolEvmLanes(lggr logger.Logger, e *cldf.Environment, state stateview.CCIPOnChainState, homeCS, feedCS uint64, laneConfig *LaneConfiguration) (cldf.Environment, error) {
```
Similar to EVM but can bypass reviewing `setupSolEvmLanes` for now as this PR will only focus on evm2evm

mateusz-sekara commented on integration-tests/load/ccip/destination_gun.go:
```diff
@@ -63,9 +93,33 @@ func NewDestinationGun(
+	}
+
+	// Fallback to any-to-any if no lane config or no sources found
+	if len(availableSources) == 0 {
```
Hmm, I think it might be better to panic/error and enforce laneConfig is set. Implicitly defaulting to any-to-any might be tricky and hard to spot when param is missing

mateusz-sekara commented on deployment/environment/crib/lane_config.go:
```diff
@@ -0,0 +1,520 @@
+)
+
+func (lc *LaneConfiguration) Validate(e *cldf.Environment) error {
+	if lc == nil {
```
ditto, it would be better to make that explicit and error when config is not present

mateusz-sekara replied:
So when I say "I need 350 lanes" it means 350 bidirectional lanes ;)

mateusz-sekara replied:
I think it would simplify the code slighlty

asoliman92 committed:
Fix lane deployment

Because we always ran it with bidirectional lanes the bug wasn't caught, we were swapping dst and src in the offRampUpdates.

Made it more parallel making around 20% better time locally while connecting lanes

asoliman92 committed:
Fix evmSources keys

We were overriding evmSources[src] instead of using different ones

mateusz-sekara commented on deployment/environment/crib/lane_config.go:
```diff
@@ -0,0 +1,512 @@
+	NumLanes *int `toml:",omitempty"`
+
+	// EnsureBidirectional - when true, ensures that for every A->B lane, there's also a B->A lane
+	EnsureBidirectional *bool `toml:",omitempty"`
```
I think this is redundant, lanes should always be bidirectional

mateusz-sekara commented on deployment/environment/crib/ccip_deployer.go:
```diff
@@ -607,42 +621,113 @@ func setupSolLinkPools(e *cldf.Environment) (cldf.Environment, error) {
 
+	// Filter lanes to only include Sol <-> EVM combinations
+	solEvmLanes := make([]LaneConfig, 0)
+	evmChainSet := make(map[uint64]bool)
```
Will it complicate things when we introduce a new non-EVM chain? I guess that would require all possible combinations?

asoliman92 replied:
It can complicate it a little bit yes. It is how it's designed though. I'll leave any Solana related code for a later PR though. This is just like a guideline for when I start introducing solana lanes afterwards.

mateusz-sekara commented on integration-tests/load/ccip/ccip_test.go:
```diff
@@ -418,15 +441,15 @@ func TestCCIPLoad_RPS(t *testing.T) {
-			t.Fail()
-		}()
-	}
+	//timeout := userOverrides.GetTimeoutDuration()
```
Remove if not needed

asoliman92 replied:
accidental push 😅

mateusz-sekara commented on integration-tests/load/ccip/destination_gun.go:
```diff
@@ -130,23 +186,48 @@ func (m *DestinationGun) Call(_ *wasp.Generator) *wasp.Response {
 func (m *DestinationGun) sendEVMSourceMessage(src uint64) error {
-	acc := m.evmSourceKeys[src]
-	r := m.state.Chains[src].Router
+	// CRITICAL: Lock the source chain to prevent nonce collisions
```
Hmm, is that a problem? Nonces are per sender per lane (source -> dest), and guns are per destination. So guns should not be racing each other, no?

asoliman92 replied:
I actually found a bug with senders that I fixed 2 hours ago. We were using same sender for a source chain across all destinations. I think the other fix should be enough. Will test again and make sure it works and then can remove this part.

asoliman92 replied:
Multi senders' fix took care of it. Thanks for pointing out. Removed from PR.

asoliman92 committed:
Ensure bidirecionality for all lanes

asoliman92 committed:
more tests

asoliman92 committed:
more tests + linting

asoliman92 commented on deployment/environment/crib/ccip_deployer.go:
```diff
@@ -745,14 +873,16 @@ func setupEVM2EVMLanes(e *cldf.Environment, state stateview.CCIPOnChainState) (c
 					IsRMNVerificationDisabled: true,
 				}
 
-				updateRouterChanges[src].OffRampUpdates[dst] = true
```
That was a bug that we never caught because we were always doing bidirectional lanes. Although now we still only do bidirectional lanes, it's more reliable.

asoliman92 commented on deployment/environment/crib/ccip_deployer.go:
```diff
@@ -745,14 +873,16 @@ func setupEVM2EVMLanes(e *cldf.Environment, state stateview.CCIPOnChainState) (c
-				updateRouterChanges[src].OnRampUpdates[dst] = true
-				mu.Lock()
+				globalUpdateRouterChanges[dst].OffRampUpdates[src] = true
+				globalUpdateRouterChanges[src].OnRampUpdates[dst] = true
```
This optimization is making lane deployment around 10% better for me locally. As I got the onRamp/offRamp changesents it's easier to run current changesets in parallel.

asoliman92 committed:
Remove mutex on sourceChains

After fixing having different senders per src->dst lanes we don't need to have the mutex in place anymore

asoliman92 committed:
linting

asoliman92 committed:
Merge branch 'develop' into asoliman/load-tests/lane-config

asoliman92 committed:
Upgrade mergo to use latest path

asoliman92 committed:
Revert "Upgrade mergo to use latest path"

This reverts commit 7efe32db763771ee408ab965feb1dfff89943d49.

asoliman92 committed:
Merge branch 'develop' into asoliman/load-tests/lane-config

0xAustinWang commented on integration-tests/load/ccip/ccip_test.go:
```diff
@@ -220,34 +226,51 @@ func TestCCIPLoad_RPS(t *testing.T) {
 			case selectors.FamilyEVM:
-				evmSourceKeys[src] = evmSenders[src][ind]
-
+				// Check if we have enough senders for this source chain
+				if len(evmSenders[src]) <= ind {
+					lggr.Errorw("Not enough EVM senders for source chain",
+						"sourceChain", src,
+						"destinationChain", cs,
+						"requiredIndex", ind,
+						"availableSenders", len(evmSenders[src]))
+					mu.Unlock()
+					continue
+				}
+				evmSourceKeys[cs][src] = evmSenders[src][ind]
```
Let's reverse this block, 
```suggestion
				if len(evmSenders[src]) > ind { 
				 evmSourceKeys[cs][src] = evmSenders[src][ind]
				}```

asoliman92 committed:
review comments
